File: 1267-remove-zero-sum-consecutive-nodes-from-linked-list/remove-zero-sum-consecutive-nodes-from-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullnode) {}
 *     ListNode(int x) : val(x), next(nullnode) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeZeroSumSublists(ListNode* head) 
    {
        unordered_map<int,ListNode*> map;
        ListNode* dummy=new ListNode(-1);
        dummy->next=head;
        map[0]=dummy;

        int pSum=0;
        ListNode* curr=head;
        while(curr)
        {
            pSum+=curr->val;
            if(map.count(pSum)>0)
            {
                ListNode* node=map[pSum]->next;
                int tempSum=pSum;
                // erasing elements in the map present between zero sum nodes.
                // dry run this example to understand [1,3,2,-3,-2,5,5,-5,1]
                while(node!=curr)         
                {
                    tempSum+=node->val;
                    map.erase(tempSum);
                    node=node->next;
                }
                map[pSum]->next=curr->next;
            }
            else
            map[pSum]=curr;
            
            curr=curr->next;
        }
        return dummy->next;
    }
};

File: 1267-remove-zero-sum-consecutive-nodes-from-linked-list/remove-zero-sum-consecutive-nodes-from-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeZeroSumSublists(ListNode head) {
                ListNode dummyHead = new ListNode(0);
        dummyHead.next = head;

        Map<Integer, ListNode> map = new HashMap<>();
        map.put(0, dummyHead);
        
        ListNode curr = head;
        int prefixSum = 0;
        while(curr != null){
            prefixSum += curr.val;
            if(map.containsKey(prefixSum)){
                ListNode ptr = map.get(prefixSum).next;
                int tempSum = prefixSum;
                // remove the prefix sum stored in the map
                while(ptr != curr){
                    tempSum += ptr.val;
                    map.remove(tempSum);
                    ptr = ptr.next;
                }
                // make connection
                map.get(prefixSum).next = curr.next;
            }else {
                map.put(prefixSum, curr);
            }
            curr = curr.next;
        }
        return dummyHead.next;
    }
}